File: solutions/python/robot-simulator/1/robot_simulator.py
EAST = 'east'
NORTH = 'north'
WEST = 'west'
SOUTH = 'south'

class Robot:
    def __init__(self, direction=NORTH, x_pos=0, y_pos=0):
        self._direction = direction
        self._coordinates = (x_pos, y_pos)

    @property
    def direction(self):
        return self._direction

    @property
    def coordinates(self):
        return self._coordinates

    def move(self, instructions):
        for instruction in instructions:
            match instruction:
                case 'L':
                    self.turn_left()
                case 'R':
                    self.turn_right()
                case 'A':
                    self.advance()

    def turn_left(self):
        directions = [NORTH, WEST, SOUTH, EAST]
        index = directions.index(self._direction)
        self._direction = directions[(index + 1) % 4]

    def turn_right(self):
        directions = [NORTH, EAST, SOUTH, WEST]
        index = directions.index(self._direction)
        self._direction = directions[(index + 1) % 4]

    def advance(self):
        x, y = self._coordinates
        match self.direction:
            case 'north':
                y += 1
            case 'east':
                x += 1
            case 'south':
                y -= 1
            case 'west':
                x -= 1
        self._coordinates = (x, y)

File: solutions/typescript/robot-simulator/1/robot-simulator.ts

export class InvalidInputError extends Error {
    constructor(message: string) {
        super();
        this.message = message || 'Invalid Input';
    }
}

type Direction = 'north' | 'east' | 'south' | 'west'
type Coordinates = [number, number]

export class Robot {
    private _bearing: Direction;
    private _coordinates: Coordinates;

    constructor() {
        this._bearing = 'north';
        this._coordinates = [0, 0];
    }

    get bearing(): Direction {
        return this._bearing;
    }

    get coordinates(): Coordinates {
        return this._coordinates;
    }

    place({ x, y, direction }: { x: number; y: number; direction: string }) {
        const d = direction;
        if (d !== 'north' && d !== 'east' && d !== 'south' && d !== 'west') {
            throw new InvalidInputError(`Invalid direction: ${direction}`);
        }
        this._coordinates = [x, y];
        this._bearing = d;
    }

    evaluate(instructions: string) {
        for (const instruction of instructions) {
            switch (instruction) {
                case 'L': this.turnLeft(); break;
                case 'R': this.turnRight(); break;
                case 'A': this.advance(); break;
                default: throw new InvalidInputError(
                    `Invalid instruction: ${instruction}`,
                );
            }
        }
    }

    private turnLeft() {
        const directions: Direction[] = ['north', 'west', 'south', 'east'];
        this._bearing = directions[(directions.indexOf(this._bearing) + 1) % 4];
    }

    private turnRight() {
        const directions: Direction[] = ['north', 'east', 'south', 'west'];
        this._bearing = directions[(directions.indexOf(this._bearing) + 1) % 4];
    }

    private advance() {
        switch (this._bearing) {
            case 'north': this._coordinates[1]++; break;
            case 'east': this._coordinates[0]++; break;
            case 'south': this._coordinates[1]--; break;
            case 'west': this._coordinates[0]--; break;
        }
    }
}
